packages/markers/src/browser/problem/problem-decorator.ts

```diff
@@ -168,3 +171,131 @@ export namespace ProblemDecorator {
+
+    @postConstruct()
+    protected init(): void {
+        this.emitter = new Emitter();
+        this.problemManager.onDidChangeMarkers(() => this.fireDidChangeDecorations((titles: Title<Widget>[]) => this.collectDecorators(titles)));
+    }
+
+    decorate(titles: Title<Widget>[]): Map<string, TreeDecoration.Data> {
+        return this.collectDecorators(titles);
+    }
+
+    get onDidChangeDecorations(): Event<(titles: Title<Widget>[]) => void> {
+        return this.emitter.event;
+    }
+
+    protected fireDidChangeDecorations(event: (titles: Title<Widget>[]) => Map<string, TabBarDecoration.Data>): void {
+        this.emitter.fire(event);
+    }
+
+    /**
+     * Collect decorators for the tabs.
+     * @returns {Map<string, TabBarDecoration.Data>} A map from the tab URI to the tab decoration data.
+     */
+    protected collectDecorators(titles: Title<Widget>[]): Map<string, TabBarDecoration.Data> {
+        const result: Map<string, Marker<Diagnostic>> = new Map();
+        const markers = this.groupMarkersByURI(this.collectMarkers());
+        for (const title of titles) {
+            const relativeUri = title.caption;
+            const root = this.workspaceService.tryGetRoots()[0];
+            const path = root.uri + '/' + relativeUri;
+            const marker = markers.get(path);
+            if (marker) {
+                result.set(relativeUri, marker);
+            }
+        }
+        const urlDecoratorMap = new Map(Array.from(result.entries())
+            .map(entry => [entry[0], this.toDecorator(entry[1])] as [string, TabBarDecoration.Data]));
+        return urlDecoratorMap;
+    }
+
+    /**
+     * Group markers by the URI of the editor they decorate.
+     * @param {Marker<Diagnostic>[]} markers - all diagnostic markers collected/
+     * @returns {Map<string, Marker<Diagnostic>>} A map from URI of the editor to its diagnostic markers.
+     */
+    protected groupMarkersByURI(markers: Marker<Diagnostic>[]): Map<string, Marker<Diagnostic>> {
+        const result: Map<string, Marker<Diagnostic>> = new Map();
+
+        for (const [uri, marker] of new Map(markers.map(m => [new URI(m.uri), m] as [URI, Marker<Diagnostic>])).entries()) {
+            const uriString = uri.toString();
+            result.set(uriString, marker);
+        }
+        return result;
+    }
+
+    /**
+     * Collect all diagnostic markers from the problem manager.
+     * @returns {Marker<Diagnostic>[]} An array of diagnostic markers.
+     */
+    protected collectMarkers(): Marker<Diagnostic>[] {
+        return Array.from(this.problemManager.getUris())
+            .map(str => new URI(str))
+            .map(uri => this.problemManager.findMarkers({ uri }))
+            .map(markers => markers.sort(this.compare.bind(this)).shift())
+            .filter(notEmpty)
+            .filter(this.filterMarker.bind(this));
+    }
+
+    protected toDecorator(marker: Marker<Diagnostic>): TabBarDecoration.Data {
```

Comment:
Some methods have documentation while others do not, it'd be good to be consistent and add documentation always.